### PR TITLE
Add ILIAS plugin

### PR DESCRIPTION
**Pull request by pawohl**
- [x] [Abide with _The Concise guide to plugin authoring_](http://guide.munin-monitoring.org/en/latest/develop/plugins/plugin-concise.html)
- [x] Make the plugin file executable (+x permission) so tests pass
- [x] check if a similar plugin exists and if it can be extended/changed instead of adding a new plugin
  There is no similar plugin in scope.
- [x] add documentation including configuration, author, license and magic markers
  Done. See code.
- [x] pick a suitable category
   cms although it's an LMS. Following the moodle sample.
- [x] use style check tools for the language of the plugin
   used pep8online
- [x] pick a suitable name and location
   ilias (name of the software)
- [x] add example graphs for the gallery
   added daily session graph
- [x] open a pull request with your new plugin or send it attached to an email to the mailing list
   pull request preferred

### CODE REVIEW

**Comment by pawohl** [comment]
[ILIAS](https://www.ilias.de/docu/goto_docu_cat_580.html) is an open source web-based learning management system (LMS).

File: plugins/ilias/ilias_ [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        cursor = self.con.cursor()
+        cursor.execute(
+            "SELECT COUNT( usr_id ) "
+            "FROM `usr_data` "
+            "WHERE last_login >= DATE_SUB( NOW( ) , INTERVAL 1 DAY )")
+        usrs = cursor.fetchone()[0]
+        print("iltotal1day.value %s" % (usrs))
+
+    def run(self):
+        cmd = ((len(sys.argv) > 1) and sys.argv[1] or None) or "execute"
```
Comment (sumpfralle) [inline plugins/ilias/ilias_]: What is the effect of `or None`?

File: plugins/ilias/ilias_ [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        function = None
+
+        if cmd == "config":
+            function = "config"
+        elif cmd == "suggest":
+            print ("sessions")
```
Comment (sumpfralle) [inline plugins/ilias/ilias_]: [PEP8](https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements) does not like whitespace between function and arguments.
May try flake8 or something similar for a style check?
(sadly this is not part of our automatic testing environment, yet)

Regarding pep8online: it does not seem to check the above issue. I reported this to the author of that service.

File: plugins/ilias/ilias [hunk 3]
```diff
@@ -0,0 +1,253 @@
+#!/usr/bin/env bash
+# Munin plugin for ILIAS
+
+: << =cut
+
+=head1 NAME
+
+ilias - Munin plugin to monitor L<ILIAS|https://ilias.de/> open source
```
[the remark this answers is not in the document]
Comment (pawohl) [reply plugins/ilias/ilias]: Done. `ilias_session`

File: plugins/ilias/ilias [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    echo "$global_attr" | sed -e 's/^  *//' -e '/^$/d'
+
+    i=0
+    while [ ${d_attr[$i,field]+ok} ]; do
+       field=${d_attr[$i,field]}
+       echo "$field.type ${d_attr[$i,type]}"
+       echo "$field.draw ${d_attr[$i,draw]}"
+       echo "$field.label ${d_attr[$i,label]:0:${label_max_length}}"
+       echo "$field.min 0"
+       ((i++)) || true
```
[the remark this answers is not in the document]
Comment (pawohl) [reply plugins/ilias/ilias]: Done. `((++i))` does the job in this case (0-based indexing).

File: plugins/ilias/ilias [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+config() {
+    local label_max_length=45
+    local i=0
+
+    # print global attributes
+    echo "$global_attr" | sed -e 's/^  *//' -e '/^$/d'
+
+    i=0
+    while [ ${d_attr[$i,field]+ok} ]; do
```
[the remark this answers is not in the document]
Comment (pawohl) [reply plugins/ilias/ilias]: I am only checking that an element was set. C.f. https://stackoverflow.com/a/13221491/2683737 -- if the element is an empty string, somebody did something wrong but checking that is beyond this plugin's scope. Using `-n`, the test always passed:

```
[ -n ${d_attr[$i,field]+ok} ] && echo "good" || echo "not good"
# always echoes "good"
```

Therefore I left this unmodified.

[the remark this answers is not in the document]
Comment (sumpfralle) [reply plugins/ilias/ilias]: > Using -n, the test always passed:

"test" without quotes is a tricky beast :)
```
[ -n ]; echo $?
[ -n "" ]; echo $?
```
`-n` would work, if you would quote the expression. Anyway - just details ...

Regarding `${foo+ok}`: this is a really nifty (and hard to read and even harder to look up in a manual) test for existence. Maybe the [other recommendation](https://stackoverflow.com/a/45385463) in the stackoverflow discussion is easier to read:
```
[[ -v $d_attr[$i,field] ]]
```

Anyway: this is a matter of taste. Just keep it, if you prefer so.

File: plugins/ilias/ilias_session [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+In order to get precise results, please ensure your MySQL server has the same
+time as your ILIAS application server. Timezone does not matter.
+
+=head1 CONFIGURATION
+
+The plugin needs the following configuration settings e.g. in
+/etc/munin/plugin-conf.d/ilias.conf
+
+    [ilias]
```
Comment (sumpfralle) [inline plugins/ilias/ilias_session]: Remained of renaming: `ilias_session`

Comment (pawohl) [reply plugins/ilias/ilias_session]: Done. Good catch. Thank you!

**Comment by sumpfralle** [comment]
Great - thank you!

**Comment by pawohl** [comment]
> Great - thank you!

@sumpfralle Thank you!

Maybe this checklist is useful to others as well? Maybe add it to the contribs guide?

```
- [ ] [Abide with _The Concise guide to plugin authoring_](http://guide.munin-monitoring.org/en/latest/develop/plugins/plugin-concise.html)
- [ ] Make the plugin file executable (+x permission) so tests pass
- [ ] check if a similar plugin exists and if it can be extended/changed instead of adding a new plugin
  ...result...
- [ ] add documentation including configuration, author, license and magic markers
  ...result...
- [ ] pick a suitable category
   ...Which one and why?...
- [ ] use style check tools for the language of the plugin
   used ...
- [ ] pick a suitable name and location
   ...Which one and why?...
- [ ] add example graphs for the gallery
   added ... graph
- [ ] open a pull request with your new plugin or send it attached to an email to the mailing list
```

**Comment by sumpfralle** [comment]
> Maybe this checklist is useful to others as well?

Yes, indeed it would be useful.
Right now the shorter version of this list is visible below the [`plugins` directory](https://github.com/munin-monitoring/contrib/tree/master/plugins#submit-a-new-plugin). It is hard to find there, I guess. Please feel welcome to enhance this list.

Additionally I think, it could be useful to prepare it as an issue template here in the github issue tracker. Please advise, if you have experience with this.

Or where else would you see this list?
(I would like to keep the number of text copies low - and I have no real opinion right now, where the "master" text should be located)

**Comment by Rillke** [comment]
> Additionally I think, it could be useful to prepare it as an issue template here in the github issue tracker. Please advise, if you have experience with this.

GitHub allows for the creation of different issue-templates and a pull request template under `Insights` -> `Community`. Creating a pull request template using the Community-Checklist, the file will be created in the root directory of the repository and use upper case letters in its file name. This might not be what you want. [The GitHub help article on *Creating a pull request template for your repository*](https://help.github.com/articles/creating-a-pull-request-template-for-your-repository/) shows all options available.

**Pull request merged**